walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,142 @@
+    long requestDuration = request.getDuration();
+    Collection<String> requestAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    ArrayList<TimeRange> validTimes = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> validTimesWOptional = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> noOverlap = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> noOverlapWOptional = new ArrayList<TimeRange>();    
+    ArrayList<TimeRange> availableTimes = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> availableTimesWOptional = new ArrayList<TimeRange>();  
+    int newStart;
+    int newEnd;
+
+    //edge case: if there are no attendees the whole day is returned        
+    if (requestAttendees.isEmpty() && optionalAttendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    //edge case: if the requested meeting is longer than a day there are no options
+    if (requestDuration >= TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();  
+    }
+    validTimes = addValidTimes(events, request);
+    validTimesWOptional = addValidTimesWithOptional(events, request, validTimes);
+    //if there are no meetings the whole day is returned 
+    if(validTimes.isEmpty() && validTimesWOptional.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);    
+    }  
+    noOverlap = removeOverlap(validTimes);
+    noOverlapWOptional = removeOverlap(validTimesWOptional);
+    availableTimesWOptional = addAvailableRanges(request, noOverlapWOptional);
+    availableTimes = addAvailableRanges(request, noOverlap);
+    //return the list of available time ranges
+    if(!availableTimesWOptional.isEmpty()) {
+      return availableTimesWOptional;
+    }
+    return availableTimes;
+  }
+  
+  //if there are multiple meetings consolidate overlapping meetings into one range of time
+  public ArrayList<TimeRange> removeOverlap(ArrayList<TimeRange> times) {
+    ArrayList<TimeRange> noOverlap = new ArrayList<TimeRange>();
+    Collections.sort(times, TimeRange.ORDER_BY_START);
+    int newStart;
+    int newEnd;
+    if(times.size() > 1){
+      for (int i = 1; i < times.size(); i ++){
+        if (times.get(i - 1).overlaps(times.get(i))){
+          newStart = Math.min(times.get(i - 1).start(), times.get(i).start());
+          newEnd = Math.max(times.get(i).end(), times.get(i - 1).end());
+          if (!noOverlap.contains(TimeRange.fromStartEnd(newStart, newEnd, false))) {
+            noOverlap.add(TimeRange.fromStartEnd(newStart, newEnd, false));
+          }
+        } else {
+          noOverlap.add(times.get(i - 1));
+          noOverlap.add(times.get(i));
+        } 
+      }      
+    } else {
+      return times;
+    }
+      return noOverlap;
+  }
+
+  //add the times for all meetings where there are mandatory attendees
+  public ArrayList<TimeRange> addValidTimes(Collection<Event> events, MeetingRequest request) {
```

Comment:
I think the term "valid times" is a little confusing, since that makes me think it's the times when people are available, but in this case it is the times when people are in meetings, right?  If so it might be better to change this to something like `getAlreadyBookedTimes` or `getUnavailableTimes` or something like that

Reply:
addressed in "address nits(formatting and variable names)"